fem/tmop_tools.cpp

```diff
@@ -444,15 +444,15 @@ double TMOPNewtonSolver::ComputeScalingFactor(const Vector &x,
       if (untangling == false && min_detT_out < 0.0)
       {
          // No untangling, and detJ got negative -- no good.
-         if (print_level >= 0)
-         { mfem::out << "Scale = " << scale << " Neg det(J) found.\n"; }
+         if (print_options.warnings)
+         { mfem::err << "Scale = " << scale << " Neg det(J) found.\n"; }
```

Comment:
This message is informational not a warning, so it makes more sense to use `mfem::out`, not `mfem::err`.

I'm not sure what is the common consensus for warnings -- print to `stdout` or to `stderr`? For example, `mfem_warning` uses `mfem::out`, not `mfem::err`. Was this discussed previously in this PR?

Basically, `print_options.warnings` triggers both "warnings" and "info" messages -- unless we want to split them as separate options.

Actually, in this particular case, this message makes most sense to be printed only when `print_options.iterations` is enabled -- without that context, these don't make much sense.

Reply:
We had a brief discussion here https://github.com/mfem/mfem/pull/2519#issuecomment-929644919 (ping @pazner). I see your point that this is informal, but I also see the benefit of having this information available in the error log. I do not have a strong opinion here and I am fine with either.

I will add the restriction later.

Reply:
Just to clarify (since my first comment may be a bit confusing): in this specific case, I think we want to have:
```c++
         if (print_options.iterations)
         { mfem::out << "Scale = " << scale << " Neg det(J) found.\n"; }
```
The discussion about where to print warning messages will apply to other places.

Reply:
Thanks for the clarification Veselin. I think the issue here is within how I defined the `warning` and `iteration` flags. I will copy and paste the current state if the docs here:
```cpp
      /** @brief If a fatal problem has been detected some context-specific
          information will be reported */
      bool errors = false;
      /** @brief If a non-fatal problem has been detected some context-specific
          information will be reported */
      bool warnings = false;
      /// Detailed information about each iteration will be reported
      bool iterations = false;
```
With this definition I think these messages should be printed if at least of "warning" or "iteration" is active. I see your point that we may miss some context-specific information in this case, but we can also fix this issue by printing more details in the case that "warning" is active and "iteration" is inactive - but I also think that this information gives us a hint that something may have gone wrong (in a possibly non-fatal way). 
So, what do you think? Should we try to fix the definition or should we keep the definition and update the logging?

Regarding what goes to the `out` stream and which information to `err` printing warnings to `our` is consistent with how warning and error logging is generally handled in the code base and should be prefered. I think the intention with printing this warnings to `err` (instead of `out`) to provide some context information in the error log, allowing to track down potential issues faster -- but I honestly do not know how error streams are usually utilized, because I almost always opt for customized logging solutions in my applications.

Reply:
These messages are not warnings -- finding negative `det(J)` in the Newton line search is a normal situation for TMOP. One of the goals of the line search is to find a step size that does not introduce negative `det(J)` in the mesh during the mesh movement.

I agree that most natural is for real warning messages is to print them to `mfem::err`. My only concern is about some of our nightly tests where output from `stdout` is mixed with output from `stderr` which is sometimes non-deterministic -- the output from `stderr` may appear later in the combined output and then it can be confusing to interpret. This may also trigger false test failures due to text differences.

For now, I think we can go with the "warnings to stderr" approach. If this becomes an issue in the nightly tests, or other situations, we can come back to this or just consider alternative solutions.

Reply:
This makes sense, thanks for the elaboration again.

If this may lead to non-determinstic problems, then I think it is preferable to print to stdout instead of stderr.

Reply:
I added some additional docs on this and changed the streams accordingly (and consistent with mfem_error/mfem_warning).

Reply:
@tzanio, @pazner, Are you okay with the latest change in a44a864: warnings go to stdout, instead of stderr? If you are, I think the PR is ready to be merged in `next`. I'll try to merge `master` here to resolve the conflicts and the tests failure.

Reply:
That change is fine by me, I don't have any strong opinions about this.

Reply:
I am okay with it as well.